File: app/lib/src/crypto/place_codec.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:sodium/sodium.dart';

import '../domain/place.dart';
import 'aul_crypto.dart';

/// Seals/opens places as a single framed+padded blob under the circle key K_c.
/// This is the ONE source of truth for the place wire format on the app: the
/// foreground geofence loader, the places view, and the editor all go through it.
///
/// The byte layout matches the web `placeCodec` (web/src/data/placeCodec.ts) and
/// the pinned cross-language vector (vectors/crypto-vectors.json → `place_framed`)
/// exactly, so a place created on the web opens on the app and vice-versa:
///
///  * plaintext  = terse JSON `{"n":name,"lat":lat,"lng":lng,"rad":radius}`
///  * padded     = libsodium ISO/IEC 7816-4 padding to a 256-byte block (hides
///                 the name length), BEFORE sealing
///  * framed     = nonce(24) || XChaCha20-Poly1305-IETF ciphertext
///  * ad         = utf8("aul-place:v1") — domain separation, so a place blob can
///                 never be opened as an SOS ("aul-sos:v1") or the circle name.
///  * ciphertext = base64(framed) — the single opaque column the server stores.
class PlaceCodec {
  PlaceCodec(this._crypto, {this.padBlock = 256});

  final AulCrypto _crypto;
  final int padBlock;

  /// Domain-separation AD (must match the web placeCodec) so a place ciphertext
  /// can't be replayed as an SOS or the circle name.
  static final _placeAd = Uint8List.fromList(utf8.encode('aul-place:v1'));

  /// Seals a place into the base64 ciphertext the server stores opaquely.
  String seal({
    required String name,
    required double lat,
    required double lng,
    required double radius,
    required SecureKey key,
  }) {
    final payload = <String, dynamic>{
      'n': name,
      'lat': lat,
      'lng': lng,
      'rad': radius,
    };
    final plain = Uint8List.fromList(utf8.encode(jsonEncode(payload)));
    final framed = _crypto.sealFramed(
      _crypto.pad(plain, padBlock),
      key,
      ad: _placeAd,
    );
    return base64.encode(framed);
  }

  /// Opens a place ciphertext, trying every key in [keyring] (rotation-safe).
  /// Returns null if no key opens it (wrong key, wrong AD, or malformed blob) —
  /// mirroring the web `openPlace`, which returns null rather than throwing.
  ///
  /// [createdBy] rides along untouched: it is server metadata, never part of the
  /// sealed blob, so it needs no key — the NAME is what stays E2EE.
  Place? open({
    required String id,
    required int version,
    required String ciphertextB64,
    required List<SecureKey> keyring,
    String? createdBy,
  }) {
    final Uint8List blob;
    try {
      blob = base64.decode(ciphertextB64);
    } catch (_) {
      return null;
    }
    for (final key in keyring) {
      try {
        final plain = _crypto.unpad(
          _crypto.openFramed(blob, key, ad: _placeAd),
          padBlock,
        );
        final m = jsonDecode(utf8.decode(plain)) as Map<String, dynamic>;
        return Place(
          id: id,
          version: version,
          name: m['n'] as String,
          lat: (m['lat'] as num).toDouble(),
          lng: (m['lng'] as num).toDouble(),
          radius: (m['rad'] as num).toDouble(),
          createdBy: createdBy,
        );
      } catch (_) {
        // try the next key
      }
    }
    return null;
  }
}

File: web/src/data/placeCodec.ts
import { fromBase64, openFramed, pad, sealFramed, toBase64, unpad } from '../crypto/aulCrypto';
import type { Place, PlaceDTO, SosDTO, SosEvent } from './types';

/// Fixed padding block for place/SOS payloads. Padding BEFORE sealing hides the
/// plaintext length so a place name ("Mom's house" vs "Gym") or an SOS message
/// can't leak via ciphertext size. MUST match the reporter app (Dart PlaceCodec
/// padBlock) and the pinned cross-language vector.
export const PAD_BLOCK = 256;

/// Domain-separation associated data: binds each ciphertext to its field type so
/// a place blob can't be replayed as an SOS (or vice-versa). Must match the app.
const PLACE_AD = new TextEncoder().encode('aul-place:v1');
const SOS_AD = new TextEncoder().encode('aul-sos:v1');

/// On-the-wire place payload — terse keys (matches the Dart PlaceCodec schema):
/// n=name, lat/lng=centre, rad=geofence radius in metres.
interface PlacePayload {
  n: string;
  lat: number;
  lng: number;
  rad: number;
}

/// Seals a place into the single framed+padded ciphertext the server stores
/// (nonce||ct, base64). The server never sees any of these fields.
export function sealPlace(
  place: { name: string; lat: number; lng: number; radius: number },
  key: Uint8Array,
): string {
  const payload: PlacePayload = { n: place.name, lat: place.lat, lng: place.lng, rad: place.radius };
  const plain = pad(new TextEncoder().encode(JSON.stringify(payload)), PAD_BLOCK);
  return toBase64(sealFramed(plain, key, PLACE_AD));
}

/// Opens a place DTO, trying every key in the circle keyring (rotation-safe).
/// Returns null if no key opens it (e.g. sealed under a key this device lacks).
///
/// `created_by` rides along untouched: it is server metadata, never part of the
/// sealed blob, so it needs no key — the NAME is what stays E2EE.
export function openPlace(dto: PlaceDTO, keyring: Uint8Array[]): Place | null {
  const blob = fromBase64(dto.ciphertext);
  for (const key of keyring) {
    try {
      const plain = unpad(openFramed(blob, key, PLACE_AD), PAD_BLOCK);
      const p = JSON.parse(new TextDecoder().decode(plain)) as PlacePayload;
      return {
        id: dto.id,
        version: dto.version,
        name: p.n,
        lat: p.lat,
        lng: p.lng,
        radius: p.rad,
        createdBy: dto.created_by ?? null,
      };
    } catch {
      /* try the next key */
    }
  }
  return null;
}

/// On-the-wire SOS payload: optional last-known location + free-text message.
interface SosPayload {
  lat?: number;
  lng?: number;
  msg?: string;
  ts: number;
}

/// Seals an SOS payload under K_c into the single framed+padded ciphertext.
export function sealSos(payload: SosPayload, key: Uint8Array): string {
  const plain = pad(new TextEncoder().encode(JSON.stringify(payload)), PAD_BLOCK);
  return toBase64(sealFramed(plain, key, SOS_AD));
}

/// Opens an SOS DTO with the keyring. If no key opens it, the alert is STILL
/// returned (decrypted:false) from server metadata — a watcher must never miss an
/// SOS just because the payload can't be read.
export function openSos(dto: SosDTO, keyring: Uint8Array[]): SosEvent {
  const meta: SosEvent = { id: dto.id, deviceId: dto.device_id, createdAt: dto.created_at, decrypted: false };
  let blob: Uint8Array;
  try {
    blob = fromBase64(dto.ciphertext);
  } catch {
    return meta; // malformed ciphertext — still surface the alert from metadata
  }
  for (const key of keyring) {
    try {
      const plain = unpad(openFramed(blob, key, SOS_AD), PAD_BLOCK);
      const p = JSON.parse(new TextDecoder().decode(plain)) as SosPayload;
      return { ...meta, lat: p.lat, lng: p.lng, message: p.msg, ts: p.ts, decrypted: true };
    } catch {
      /* try the next key */
    }
  }
  return meta;
}
